carddav/server.go

```diff
@@ -682,11 +684,42 @@ func (b *backend) Put(r *http.Request) (*internal.Href, error) {
 }
 
 func (b *backend) Delete(r *http.Request) error {
-	return b.Backend.DeleteAddressObject(r.Context(), r.URL.Path)
+	switch b.resourceTypeAtPath(r.URL.Path) {
+	case resourceTypeAddressBook:
+		return b.Backend.DeleteAddressBook(r.Context(), r.URL.Path)
+	case resourceTypeAddressObject:
+		return b.Backend.DeleteAddressObject(r.Context(), r.URL.Path)
+	}
+	return internal.HTTPErrorf(http.StatusForbidden, "carddav: cannot delete resource at given location")
 }
 
 func (b *backend) Mkcol(r *http.Request) error {
-	return internal.HTTPErrorf(http.StatusForbidden, "carddav: address book creation unsupported")
+	if b.resourceTypeAtPath(r.URL.Path) != resourceTypeAddressBook {
+		return internal.HTTPErrorf(http.StatusForbidden, "carddav: address book creation not allowed at given location")
+	}
+
+	ab := AddressBook{
+		Path: r.URL.Path,
+	}
+
+	// Check if a request body was sent
+	var buf []byte
+	_, err := r.Body.Read(buf)
```

Comment:
Hm, is this a reliable way to check for empty request body?

It seems like it might be:

> If len(p) == 0, Read should always return n == 0. It may return a non-nil error if some error condition is known, such as EOF. 

Nit: it would be simpler to `Read(nil)`.

Ref https://github.com/emersion/go-webdav/commit/b043bbd9657b2c47250648db9889d78295f7da27

Reply:
> Hm, is this a reliable way to check for empty request body?
> 
> It seems like it might be:
> 
> > If len(p) == 0, Read should always return n == 0. It may return a non-nil error if some error condition is known, such as EOF.
> 
> Nit: it would be simpler to `Read(nil)`.

So, I am hoping it is. If there is anything better, I am certainly open.

Also, it is up for discussion if we want to support this at all. A MKCOL with an empty body is valid according to [WebDAV](https://datatracker.ietf.org/doc/html/rfc4918#section-9.3), but I am honestly not sure if any CardDAV/CalDAV clients actually do it. Still, I thought it might be nice to at least not prevent it from working.